zebra-chain/src/types.rs

```diff
@@ -15,26 +17,67 @@ use crate::serialization::{
     ReadZcashExt, SerializationError, WriteZcashExt, ZcashDeserialize, ZcashSerialize,
 };
 
-/// A 4-byte checksum using truncated double-SHA256 (two rounds of SHA256).
-#[derive(Copy, Clone, Eq, PartialEq)]
-pub struct Sha256dChecksum(pub [u8; 4]);
+/// A runtime validated type for representing amounts of zatoshis
+#[derive(Debug, Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+pub struct Amount<C = NegativeAllowed>(i64, PhantomData<C>);
 
-impl<'a> From<&'a [u8]> for Sha256dChecksum {
-    fn from(bytes: &'a [u8]) -> Self {
-        use sha2::{Digest, Sha256};
-        let hash1 = Sha256::digest(bytes);
-        let hash2 = Sha256::digest(&hash1);
-        let mut checksum = [0u8; 4];
-        checksum[0..4].copy_from_slice(&hash2[0..4]);
-        Self(checksum)
+#[cfg(test)]
+impl<C> Arbitrary for Amount<C>
+where
+    C: AmountConstraint + fmt::Debug,
+{
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        C::valid_range().prop_map(|v| Self(v, PhantomData)).boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+impl<C> From<Amount<C>> for i64 {
+    fn from(amount: Amount<C>) -> Self {
+        amount.0
     }
 }
 
-impl fmt::Debug for Sha256dChecksum {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_tuple("Sha256dChecksum")
-            .field(&hex::encode(&self.0))
-            .finish()
+///
+pub trait AmountConstraint {
+    ///
+    fn valid_range() -> Range<i64>;
+
+    ///
+    fn validate(value: i64) -> Result<i64, &'static str> {
+        let range = Self::valid_range();
+
+        if value < range.start {
+            Err("Amount's value is less than the start of its range")
+        } else if value > range.end {
+            Err("Amount's value is greater than the end of its range")
```

Comment:
Issue: `Range` is lower-inclusive and upper-exclusive, but this code excludes both bounds.

Let's use `range.contains()` instead:
```suggestion
        if !range.contains(value) {
            Err("Amount's value is outside its valid range")
```

Reply:
Question: How can our tests catch issues like this?

Should the arbitrary tests have caught this issue?
Do we need more specific tests?

Reply:
Hmm, I wanted to try to get more specific error messages here but I can see the danger...

> Question: How can our tests catch issues like this?

:grimacing: not sure, being careful?

Reply:
Ideally, `Arbitrary` should test bounds as part of its early tests. But maybe it's not designed to do that.